### PR TITLE
feat(`crosschain`): add rate limiter flags singleton state variable

### PR DESCRIPTION
**Pull request by lumtis**
# Description

This self contained object will contains all information for rate limiting

- Add message UpdateRateLimiterFlags
- Add query
- Add variable to genesis

```
message RateLimiterFlags {
  bool enabled = 1;

  // window in blocks
  int64 window = 2;

  // rate in azeta per block
  string rate = 3 [
    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Uint",
    (gogoproto.nullable) = false
  ];

  // conversion in azeta per token
  repeated Conversion conversions = 4 [(gogoproto.nullable) = false];
}

message Conversion {
  string zrc20 = 1;
  string rate = 2 [
    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
    (gogoproto.nullable) = false
  ];
}

```

Closes: https://github.com/zeta-chain/node/issues/2041

### CODE REVIEW

File: x/crosschain/types/rate_limiter_flags.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Validate checks that the RateLimiterFlags is valid
+func (r RateLimiterFlags) Validate() error {
+	// window must not be negative
+	if r.Window < 0 {
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/types/rate_limiter_flags.go:12]: Block height are using int64, so likely the service using this query will compare this value with a int64

[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/types/rate_limiter_flags.go:12]: this is okay as I have to compare it with height which is also `int64`

**Review by ws4charlie — APPROVED** [review]

File: x/crosschain/types/codec.go [hunk 2]
```diff
@@ -32,6 +33,7 @@ func RegisterInterfaces(registry cdctypes.InterfaceRegistry) {
 		&MsgMigrateTssFunds{},
 		&MsgUpdateTssAddress{},
 		&MsgAbortStuckCCTX{},
+		&MsgUpdateRateLimiterFlags{},
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/types/codec.go:36]: What do you mean, `RegisterInterfaces ` called in newCodec

File: x/crosschain/types/rate_limiter_flags.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		return fmt.Errorf("window must be positive: %d", r.Window)
+	}
+
+	seen := make(map[string]bool)
+	for _, conversion := range r.Conversions {
+		// check no duplicated conversion
+		if _, ok := seen[conversion.Zrc20]; ok {
+			return fmt.Errorf("duplicated conversion: %s", conversion.Zrc20)
+		}
+		seen[conversion.Zrc20] = true
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/types/rate_limiter_flags.go:22]: Actually better we can check if it is an Ethereum address

**Pull request merged**